### DISCUSSION TITLE
Shopware 6.5.4.0

### DISCUSSION THREAD

**Discussion by shyim** [shopware/platform#3254] [Announcements]
Hey 👋 ,

We released Shopware 6.5.4.0 today. Here are some developer highlights:

## Added generics to EntityRepository and EntitySearchResult

We added generics to EntityRepository and EntitySearchResult. With this feature, it is more comfortable to work with your IDE (better auto-completion), but you also achieve better type safety when working with tools like PHPStan or Psalm.

Example:
```php
    /**
     * @param EntityRepository<CustomerAddressCollection> $addressRepository
     */
    public function __construct(
        private readonly EntityRepository $addressRepository,
        private readonly EventDispatcherInterface $eventDispatcher
    ) {
    }
```

any further call to `search` and `getEntities` will return a `CustomerAddressCollection`.

## Fixed extending document templates with sw_extends

Finally, it's possible to extend the document partial templates like the header on any document type instead of overriding every document type once. [NEXT-26513](https://issues.shopware.com/issues/NEXT-26513)

## Added configuration to disable fine-grained caches

A configuration to disable Shopware fine-grained caches was added. With this, the whole cache instead of just config changes can be cleared and hence, may increase the performance massively. Please also see the [documentation about disabling fine-grained caching](https://developer.shopware.com/docs/guides/hosting/performance/performance-tweaks#disable-fine-grained-caching) to learn how to use this feature.

## Trusted proxies now configurable with normal Symfony settings

How to configure Shopware to work behind a load balancer or a reverse proxy? Trusted Proxies can now be configured utilizing the [regular Symfony configuration](https://symfony.com/doc/current/deployment/proxies.html). This feature now works out of the box in Shopware, as it was not easy before to configure. 😉

## Loosen composer constraints

We loosened constraints in the composer to allow the installation of minor updates of some composer packages. This makes it easier to install, for example, specific versions of security updates or 3rd party packages. See this [commit](https://github.com/shopware/platform/commit/18383ed409) for more information.

## Logging of extensions

Monolog instances of extensions consider now the project-wide configured Monolog handler.

## And many more things
- [Monolog instances of extensions consider now the project-wide configured Monolog handler.](https://github.com/shopware/platform/commit/4faeb1a6f0bc2b8d4064e53c699434a49b51fdb5)
- [public/asset-manifest.json is now stored outside of public directory](https://github.com/shopware/platform/commit/f3184330a8e07db3e32b8ae1d60b01bb943564dc)
- [Doctrine profiler bridge is now only active in dev mode](https://github.com/shopware/platform/commit/f230974c863287876edb2d8d12fa646398922d0c)
- [Multilingual Elasticsearch indices, create one Elasticsearch per entity. Improves indexing performance. Behind Feature flag until 6.6 `ES_MULTILINGUAL_INDEX` ](https://github.com/shopware/platform/commit/59a6a76d30d17c26fc8c11b50a41d80c9a0f30c0)
- [Salutation is now optional](https://github.com/shopware/platform/commit/e561c3f199e07dbff6bdfd0b0a586c288b6c10c0)
- [Apps installed with composer can be now installed with `app:install`](https://github.com/shopware/platform/commit/756cb788d83c5d58dcc080d083f706b6c7944745)
- [Admin Worker now respects the poll interval](https://github.com/shopware/platform/commit/b276b5213158fe0dea3fa00313b948aa1c9d70b2)
- [Recurring Payment Handler](https://github.com/shopware/platform/commit/c8a1e2eb00f517a30ebe7103f8026a7ede8f33b6)
- [CacheInvalidationSubscriber can be now manipulated](https://github.com/shopware/platform/commit/4ee545ffc033d8b9c63028b377aa46588544e9b4)
- Adjusted various log levels to be more consistent 

Check out our blog post: https://www.shopware.com/en/news/shopware-6-release-news-august-2023/
Compare: https://github.com/shopware/platform/compare/v6.5.3.3...v6.5.4.0

**Comment by ioanok** [comment]
The changelog says that [NEXT-25804](https://github.com/shopware/platform/blob/v6.5.4.0/changelog/release-6-5-4-0/2023-07-11-duplicated-theme-preview.md) is fixed, but this is not true. The issue is generated by `theme:refresh` or` system:update:finish` commands.

**Comment by ssltg** [comment]
Hi @ioanok ,
checked it:
the issue is fixed for new setups. If your media database is once corrupt with generated duplicates like Logo(1), it cannot be fixed automatically. You have to go directly to the database and delete all the duplicates (as they may be still connected with the theme in the theme_media table, they cannot be deleted in the admin).
 
![grafik](https://github.com/shopware/platform/assets/45385287/deff8c29-bda3-42bb-af56-792bedd27936)
And ensure the id in the theme table is the one of the initial image (without (x) suffix )
![grafik](https://github.com/shopware/platform/assets/45385287/d0e8e7b3-7432-4e97-864b-08b06999468f)

![grafik](https://github.com/shopware/platform/assets/45385287/3fa8652e-afd0-4a21-a62b-42a9d4f4408e)

If this not fixes your setup, please provide a detailed reproducing setup.

greets 👍

**Reply by ioanok** [reply]
Thank you for clarifications, the issue is solved now.

**Reply by 7underlines** [reply]
We need to migrate a shop where the default images defaultThemePreview and favicon have been deleted, leaving only favicon_(1) and defaultThemePreview(1) in existence. I've cleaned the theme_media table and removed all duplicates. However, during the upgrade process, I still encounter the following error:
`[Shopware\Core\Content\Media\Exception\DuplicatedMediaFileNameException]  
  A file with the name "defaultThemePreview_(1).jpg" already exists.`

**Reply by ssltg** [reply]
You can try to rename the file itself in the directory and in the database by removing the suffix and then try again.

**Comment by wannevancamp** [comment]
Since 6.5.4.0 our custom `src/Resources/app/storefront/webpack.config.js` breaks. After building the frontend with `bin/build-storefront.sh` the following error is shown: `Module not found: Error: Can't resolve 'xyz.js' in ...`

The bug is introduced in: https://github.com/shopware/platform/commit/45054ab6

**Reply by shyim** [reply]
Is it possible that you provide us the extension where the error occurs?

**Reply by wannevancamp** [reply]
I've sent you the zip via PM

**Reply by tobiasberge** [reply]
Hi @wannevancamp, thank you for providing your plugin. I tested to revert the change you mentioned but I still got the "Module not found" error. I think it is related to this change here: https://github.com/shopware/platform/commit/aed307329a75eadef1b83548ae0e2db494db850d.

The error only happens for the packages which have a `.js` suffix. For example the alias `instantsearch.js`. All other package aliases seem to work just fine. We have introduced a new `extensionAlias` which now conflicts with a regular alias if it also has a `.js` suffix.

The `extensionAlias` was added for migrating JS files to TypeScript gradually while preventing breaking changes (ironically) in case somebody imports a file with explicit `.js` suffix and the file is changed to `.ts` suffix.

```js
extensionAlias: {
    '.js': ['.ts', '.js'],
},
```
I'm afraid we'll have to revert this and find another solution for it because it can break existing plugins which use an alias for package names with `.js` stuffix.

In the meantime you can fix by resolving all node_modules instead of explicit aliases for each package. Imo this is a better solution in general because it is much less maintenance when having multiple packages. In the docs we currently propose to use an alias, I made a PR to change this in the future: https://github.com/shopware/docs/pull/1077

```diff
module.exports = (params) => {
    return {
        resolve: {
+            modules: [
+                `${params.basePath}/Resources/app/storefront/node_modules`,
+            ],
-            alias: {
-                'algoliasearch/lite': resolve(
-                    join(__dirname, '..', 'node_modules', 'algoliasearch/lite')
-                ),
-                'instantsearch.js': resolve(
-                    join(__dirname, '..', 'node_modules', 'instantsearch.js')
-                ),
-                'algoliasearch-helper': resolve(
-                    join(__dirname, '..', 'node_modules', 'algoliasearch-helper')
-                ),
-                'hogan.js': resolve(
-                    join(__dirname, '..', 'node_modules', 'hogan.js')
-                ),
-                '@algolia/autocomplete-core': resolve(
-                    join(__dirname, '..', 'node_modules', '@algolia/autocomplete-core')
-                ),
-                // etc...
-            }
        }
    };
}
```

**Reply by wannevancamp** [reply]
@tobiasberge , thanks for looking into this issue. I can confirm that this solution works.

**Comment by JoshuaBehrens** [comment]
I could not find anything about HTTP status codes in the BC promise. The exception to CONTENT__MEDIA_DUPLICATED_FILE_NAME has been changed from 500 to 409 with NEXT-26928. It was sort-of breaking for a piece of code for me. Are there plans to integrate the HTTP status code to the BC promise?
